server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/MergedPointRangeQuery.java

```diff
@@ -0,0 +1,207 @@
+        for (int dim = 0; dim < numDims; dim++) {
+            int offset = dim * bytesPerDim;
+            if (compareUnsigned(lower, offset, offset + bytesPerDim, upper, offset, offset + bytesPerDim) > 0) {
+                return new MatchNoDocsQuery("disjoint ranges");
+            }
+        }
+        // Otherwise on single valued segments we can only match docs the match the UNION of the two ranges.
+        return new PointRangeQuery(field, lower, upper, numDims) {
+            @Override
+            protected String toString(int dimension, byte[] value) {
+                // Stolen from Lucene's Binary range query. It'd be best to delegate, but the method isn't visible.
+                StringBuilder sb = new StringBuilder();
+                sb.append("(");
+                for (int i = 0; i < value.length; i++) {
+                    if (i > 0) {
+                        sb.append(' ');
+                    }
+                    sb.append(Integer.toHexString(value[i] & 0xFF));
+                }
+                sb.append(')');
+                return sb.toString();
+            }
+        };
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        return new ConstantScoreWeight(this, boost) {
+            Weight multiValuedSegmentWeight;
+            Weight singleValuedSegmentWeight;
+
+            @Override
+            public boolean isCacheable(LeafReaderContext ctx) {
+                return true;
+            }
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                ScorerSupplier scorerSupplier = scorerSupplier(context);
+                if (scorerSupplier == null) {
+                    return null;
+                }
+                return scorerSupplier.get(Long.MAX_VALUE);
+            }
+
+            @Override
+            public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOException {
+                /*
+                 * If we're sure docs only have a single value for the field
+                 * we can pick the most compact bounds. If there are multiple values
+                 * for the field we have to run the boolean query.
+                 */
+                PointValues points = context.reader().getPointValues(field);
+                if (points == null) {
+                    return null;
+                }
+                if (points.size() == points.getDocCount()) {
+                    // Each doc that has points has exactly one point.
+                    return singleValuedSegmentWeight().scorerSupplier(context);
+                }
+                return multiValuedSegmentWeight().scorerSupplier(context);
+            }
+
+            @Override
+            public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
+                PointValues points = context.reader().getPointValues(field);
```

Comment:
My feeling is that we do not need to implement this method. Note that queries are used in a filter context so we do not need to worry about scores?

Reply:
Spoke with Nik offline and even though at the moment there is probably no benefit in implementing this method, in a future it can be improved on the Lucene side, so this implementation will take advance of it.